### PR TITLE
feat(thor): `thor.costmatrix.dijkstra_distance` config option for disabling A* heuristic up to a distance threshold

### PR DESCRIPTION
**Pull request by chrstnbwnkl**
## The Problem

I have seen and investigated a lot of issues with the CostMatrix A* implementation that are caused by how the expansions are guided by the A* heuristic. Especially when using `CostMatrix` with time dependent requests, the best path found for a given source/target pair heavily depends on the overall location constellation. 

## Proposed solution 

The obvious solution would be to simply scratch the A* implementation, but I don't think we have to take such drastic measures. The inconsistency of results affects maybe a few percent of results, so A* works fine in most cases, and is way faster. Instead, I propose to introduce a config parameter that lets the user set a path distance threshold within which an expansion tree's A* heuristic is always 0, making the expansion effectively work like Dijkstra, only using the distance based A* heuristic beyond that threshold. 

## Using it on real tiles

I used a fixed source/target pair and added wildly different source and target locations to it. Ideally, the result of that fixed source/target pair should always be the same regarding cost, time, and shape, regardless of what other locations are in the request. 

With the current A* implementation on 100 requests, using a time-dependent request on tiles with historical traffic, I get 3  different results for this fixed location pair. 

The more I increase the new `dijkstra_distance` threshold, the more uniform the results: 25 km gives just two different results that are _almost_ identical in outcome, and **250km (covering the entire test area) gives fully uniform results**. 

## Performance implications 

Of course there's a performance penalty to pay. I ran the 100 requests for 0, 25 and 250 km and measured performance on a single threaded valhalla_service with a warm tile cache: 

| setting | req_total | median | p95 |
|---|---:|---:|---:|
| dijkstra_distance=0 | 114.1 | 1.115 | 2.149 |
| dijkstra_distance=25000 | 122.9 | 1.218 | 2.271 |
| dijkstra_distance=250000 | 136.6 | 1.356 | 2.527 |

As in the past, I would like to advocate for correctness over performance, but in this case, we can keep the default to 0, since its a small correctness gain over a ~20% performance penalty.

### CODE REVIEW

File: src/thor/costmatrix.cc [hunk 1]
```diff
@@ -1425,8 +1431,11 @@ std::string CostMatrix::RecostFormPath(GraphReader& graphreader,
 }
 
 template <const MatrixExpansionType expansion_direction, const bool FORWARD>
-float CostMatrix::GetAstarHeuristic(const uint32_t loc_idx, const PointLL& ll) const {
-  if (locs_status_[FORWARD][loc_idx].unfound_connections.empty()) {
+float CostMatrix::GetAstarHeuristic(const uint32_t loc_idx,
+                                    const PointLL& ll,
+                                    const uint32_t path_distance) const {
+  if (path_distance < dijkstra_distance_ ||
+      locs_status_[FORWARD][loc_idx].unfound_connections.empty()) {
```
Comment (chrstnbwnkl) [inline src/thor/costmatrix.cc:1443]: this is the crucial part: if the path distance to the end of the reached edge is smaller than the config value, return a zero heuristic. This will be added to the path cost to form the sort cost.

**Comment by nilsnolde** [comment]
is "time dependent queries" with traffic (live and/or historical)? did you already try to investigate what the heuristic for given pairs is doing in those requests? is it even still admissible? 

we do have (at least inferred) heuristic problems even in our routing A*'s. we never properly investigated, I always thought most of it is coming from tiny heuristic inadmissibilities, which might be related to our whole hierarchy/shortcut model and algos as well. it's certainly complex to even plan out how to investigate properly.

I feel like this is an attempt to try and paper over such problems. we have quite a few of those in our codebase already, I'm not arguing to _not_ merge this.

I'm more interested in the discussion and possibly plan some guided AI problem diagnosis with a structured test plan. lots of work, but definitely worth it IMO. so far I mostly have (very) educated guesses and historical experience, also in other valhalla forks. would be great to talk about face to face some time.

**Comment by chrstnbwnkl** [comment]
It's a good point to think about heuristic admissibility again. I know the top speed for truck which we use to compute the heuristic is fairly low, so it's not impossible that it plays a role here. Let me check again and make sure discrepancies are _not_ caused by an inadmissible heuristic.

**Comment by nilsnolde** [comment]
I was saying this on slack, but maybe the broader audience has an opinion too

> the co-location problems are really annoying. it feels like bidir A* matrix algos just exponentially makes our heuristic & hierarchy problems worse wrt optimality. routing isn't perfect, but very verkraftbar.

**Comment by chrstnbwnkl** [comment]
Ran a big suite of test matrices to check what our A* heuristic is doing. Temporarily added `astar_heuristic` and `raw_cost` (ie pre-recost) members to the matrix json output that  and compared the two across thousands of pairs (rural, urban, short, long, auto, truck). The suite ran on the same tiles I used for the above Dijkstra vs A* comparison. I could not find any case where the heuristic was inadmissible (only in some edge cases where origin and destination were slightly different but snapped to the same edge). 


There are still some untested paths of course (didn't test all permutations of all costing parameters of course), but at least it seems that the co-location problems are not caused by inadmissible A* heuristics.

**Review by nilsnolde — DISMISSED** [review]
